### PR TITLE
Chisel

### PR DESCRIPTION
**Pull request by gambol99**
- adding some work-in-progess docs for the chisel service pattern

### CODE REVIEW

File: how-to-docs/chisel.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
+#
+## Providing Access to Third Party Services (WIP)
+#
+
+> *The Problem*: we want to provides services running in ACP access to the third party services as well as the ability to have user-based access controls. At present network access in ACP is provided via Calico, but this becomes redundent when the traffic egresses the cluster. So simply peering networks together either through VPC peering or VPN connections doesn't provide the controls we want. We could rely on user-authentication on third-party service but not all services are authenticated (take POISE) and beyond that peering networks provide's no means of auditing traffic that is traversing the bridged networks.
```
Comment (KashifSaadat) [inline how-to-docs/chisel.md]: `redundent` -> `redundant`

Comment (KashifSaadat) [reply how-to-docs/chisel.md]: `provide's` -> `provides`

File: how-to-docs/certificates.md [hunk 2]
```diff
@@ -147,7 +147,7 @@ spec:
...
 
-#### **Note: the HOSTED domains are linked to the cluster i.e. notprod.acp.homeoffice.gov.uk is in notprod and prod is prod.acp.homeoffice.gov.uk etc etc. So if the domain you are requesting for in your ingress IS NOT the specific cluster domain YOU NEED to use the http challenge and update the DNS first to point our external ingress by default DNS points to the internal ingress**
+#### **Note: the HOSTED domains are linked to the cluster i.e. notprod.acp.homeoffice.gov.uk is in notprod and prod is prod.acp.homeoffice.gov.uk etc etc. So if the domain you are requesting for in your ingress IS NOT the specific cluster domain YOU NEED to use the http challenge and update the DNS first to point our external ingress by default DNS points to the internal ingress. Note, if you are placing your ingress behind the VPN i.e. using the ingress-class `ingress-internal` the HTTP challenge will not work, please contact the ACP team.**
```
Comment (KashifSaadat) [inline how-to-docs/certificates.md]: This needs rewording, as in no-one should have `(not)prod.acp.homeoffice.gov.uk` as a domain whitelist for their namespace and should ALWAYS be on their own domain.

File: how-to-docs/chisel.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+One pattern we are exploring is the use of a proxy cluster with an authenticated side-kick to route traffic and provide end-to-end encryption. Both ACP Notprod and Prod are peered to an respective proxy cluster which is running a [Chisel](https://github.com/jpillora/chisel) server. Below is rough idea of how the chisel service works.
+
+![alt text](https://github.com/UKHomeOffice/application-container-platform/blob/master/how-to-docs/pics/chisel.png "Chisel")
+
+The workflow for this is as follows, note the following example is assuming we have peered with a network in the proxy cluster which is exposing x services.
+
+* A request via BAU the provisioning of a service on the Chisel server.
+* Once done user is provided credentials for service.
+* You add into your deployment a chisel container running in client mode and add the configuration as describe to route the traffic. In regard to DNS and hostname's, kubernetes pods permit the user to add host entries into the container DNS, enabling you to override.
```
Comment (KashifSaadat) [inline how-to-docs/chisel.md]: `as describe` -> `as described`
`DNS and hostname's` -> `DNS and hostnames`

File: how-to-docs/chisel.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+* The traffic is picked up, encrypted using an over an ssh tunnel and pushed to the Chisel server where the user credentials evaluated. Assuming everything is ok the traffic is then proxied on to destination.
```
Comment (KashifSaadat) [inline how-to-docs/chisel.md]: `encrypted using an over an ssh tunnel` - wording
`where the user credentials evaluated.` -> `where the user credentials are evaluated.`

File: how-to-docs/chisel.md (hunk 1 above)
Comment (marcinc) [inline how-to-docs/chisel.md]: we want to `provide` services...  
tyops: `redundent` -> `redundant`, `provide's` -> `provides`

File: how-to-docs/certificates.md (hunk 2 above)
Comment (marcinc) [inline how-to-docs/certificates.md]: `etc etc` -> `etc`

File: how-to-docs/chisel.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        # this is saying listen on port 10443 and route all traffic to another-service.example.com:443 endpoint
+        - 127.0.0.1:10443:another-service.example.com:443
+        - 127.0.0.1:10444:example-api.internal.homeoffice.gov.uk:443
+        volumeMounts:
+        - name: bundle
+          mountPath: /etc/ssl/certs
+          readOnly: true
+```
+
+The above embeds the sidekick into the Pod and requests the client to listen on localhost:10443 and 10444 to redirect traffic via the Chisel service. The one annoying here is the port requirements, placing things on different ports, but unfortunately this is required. You should be able to call the service via `curl https://another-service.example.com:10443` at this point.
```
Comment (KashifSaadat) [inline how-to-docs/chisel.md]: `The one annoying here` -> `The one annoying point here`?

File: how-to-docs/chisel.md (hunk 3 above)
Comment (marcinc) [inline how-to-docs/chisel.md]: `as described`

File: how-to-docs/chisel.md (hunk 4 above)
Comment (marcinc) [inline how-to-docs/chisel.md]: * encrypted over an ssh tunnel and ... ? 
* where the user credentials `are` evaluated.

**Comment by gambol99** [comment]
@KashifSaadat .. fixed up things highlighted and cleaned up some spelling mistakes

File: how-to-docs/chisel.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+The workflow for this is as follows, note the following example is assuming we have peered with a network in the proxy cluster which is exposing x services.
+
+* A request via BAU the provisioning of a service on the Chisel server.
+* Once done user is provided credentials for service.
+* You add into your deployment a chisel container running in client mode and add the configuration as described to route the traffic. In regard to DNS and hostnames, kubernetes pods permit the user to add host entries into the container DNS, enabling you to override.
+* The traffic is picked up, encrypted over an ssh tunnel and pushed to the Chisel server where the user credentials are evaluated. Assuming everything is ok the traffic is then proxied on to destination.
+
+### A Working Example
+
+We have a two services called `example-api.internal.homeoffice.gov.uk` and `another-service.example.com` and we wish to consume the API from the pods. Lets assume the service as already been provisioned on the Chisel server and we have the credentials at hand.
```
Comment (marcinc) [inline how-to-docs/chisel.md]: `as already` -> `has already`

File: how-to-docs/certificates.md [hunk 7]
```diff
@@ -145,9 +145,9 @@ spec:
...
+The above example requires that the DNS zone of the ingress hostname you are adding is located in the AWS account that kube-cert-manager resides, i.e. it has the ability to update records on your behalf. Assuming this is not the case you will need to use a HTTP challenge rather than the default DNS one. You will need to change your DNS hostname to a CNAME of ingress-external / ingress-internal ENVIRONMENT.acp.homeoffice.gov.uk i.e assuming notprod environment mysite.domain.com -> ingress-external.notprod.acp.homeoffice.gov.uk
 
-#### **Note: the HOSTED domains are linked to the cluster i.e. notprod.acp.homeoffice.gov.uk is in notprod and prod is prod.acp.homeoffice.gov.uk etc etc. So if the domain you are requesting for in your ingress IS NOT the specific cluster domain YOU NEED to use the http challenge and update the DNS first to point our external ingress by default DNS points to the internal ingress**
+#### **Note: if you are placing your ingress behind the VPN i.e. using the ingress-class `ingress-internal`, the HTTP challenge will not work, please contact the ACP team as you will need to move or delegate you domain to Route53**
```
Comment (KashifSaadat) [inline how-to-docs/certificates.md]: `you domain to Route53` -> `your domain to Route53`

Maybe for a different PR, but we should provide a set of ingress examples for the different scenarios (internal domain, external domain, managed / not managed by Route53 etc).

File: how-to-docs/chisel.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
+#
+## Providing Access to Third Party Services (WIP)
+#
+
+> *The Problem*: we want to provides services running in ACP access to the third party services as well as the ability to have user-based access controls. At present network access in ACP is provided via Calico, but this becomes redundant when the traffic egresses the cluster. So simply peering networks together either through VPC peering or VPN connections doesn't provide the controls we want. We could rely on user-authentication on third-party service but not all services are authenticated (take POISE) and beyond that peering networks provides no means of auditing traffic that is traversing the bridged networks.
```
Comment (KashifSaadat) [inline how-to-docs/chisel.md]: `we want to provides services` -> `we want to provide services`

File: how-to-docs/certificates.md (hunk 7 above)
Comment (oyelekci) [reply how-to-docs/certificates.md]: Agree - I'm confused too. There is such a ticket btw but I've been told to wait till the ingress placeholder ticket's decided on - https://github.com/UKHomeOffice/application-container-platform-board/issues/969

**Pull request merged**